internal/app/export/service.go

```diff
@@ -4,62 +4,294 @@
 	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"sync/atomic"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/mainflux/export/internal/pkg/messages"
 	"github.com/mainflux/export/internal/pkg/routes"
 	"github.com/mainflux/export/internal/pkg/routes/mfx"
 	"github.com/mainflux/export/pkg/config"
-	log "github.com/mainflux/mainflux/logger"
+	logger "github.com/mainflux/mainflux/logger"
 	nats "github.com/nats-io/nats.go"
 )
 
 type Service interface {
-	Start(queue string, nc *nats.Conn)
+	Start(queue string) error
+	Subscribe(topic string, nc *nats.Conn)
 }
 
 var _ Service = (*exporter)(nil)
 
 type exporter struct {
-	Mqtt   mqtt.Client
-	Logger log.Logger
-	Cfg    config.Config
-	Routes []routes.Route
+	ID        string
+	MQTT      mqtt.Client
+	Logger    logger.Logger
+	Cfg       config.Config
+	Consumers map[string]routes.Route
+	Cache     messages.Cache
+	connected chan bool
+	status    uint32
+	sync.RWMutex
 }
 
+const (
+	exportGroup = "export-group"
+	count       = 100
+
+	disconnected uint32 = iota
+	connected
+
+	NatsSub = "export"
+	NatsAll = ">"
+)
+
+var errNoRoutesConfigured = errors.New("No routes configured")
+
 // New create new instance of export service
-func New(mqtt mqtt.Client, c config.Config, logger log.Logger) Service {
-	routes := make([]routes.Route, 0)
+func New(c config.Config, cache messages.Cache, l logger.Logger) Service {
+	routes := make(map[string]routes.Route)
+	id := fmt.Sprintf("export-%s", c.MQTT.Username)
+
 	e := exporter{
-		Mqtt:   mqtt,
-		Logger: logger,
-		Cfg:    c,
-		Routes: routes,
+		ID:        id,
+		Logger:    l,
+		Cfg:       c,
+		Consumers: routes,
+		Cache:     cache,
+		connected: make(chan bool, 1),
 	}
-
+	client, err := e.mqttConnect(c, l)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	e.MQTT = client
 	return &e
 }
 
-// Start method starts consuming messages received from NATS
-// and makes routes according to the configuration file.
-// Routes export messages to mqtt.
-func (e *exporter) Start(queue string, nc *nats.Conn) {
+// Start method loads route configuration
+func (e *exporter) Start(queue string) error {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
+		natsTopic := fmt.Sprintf("%s.%s", NatsSub, r.NatsTopic)
 		switch r.Type {
 		case "mfx":
-			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = mfx.NewRoute(natsTopic, r.MqttTopic, r.SubTopic)
 		default:
-			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = routes.NewRoute(natsTopic, r.MqttTopic, r.SubTopic)
+		}
+		if !e.validateSubject(route.NatsTopic()) {
+			continue
+		}
+		e.Consumers[route.NatsTopic()] = route
+		if e.Cache != nil {
+			g, err := e.Cache.GroupCreate(r.NatsTopic, exportGroup)
+			if err != nil {
+				e.Logger.Error(fmt.Sprintf("Failed to create stream group %s", err.Error()))
+			}
+			e.Logger.Info(fmt.Sprintf("Stream group %s created %s", r.NatsTopic, g))
+		}
+
+	}
+	if len(e.Consumers) == 0 {
```

Comment:
Can you move this check to the top?

Reply:
I'm checking here if routes are configured, if there are no routes configured then export service will do nothing